### PR TITLE
HomeWork: Part1 + Part2 (Aleksandr_Saunin@epam.com)

### PR DESCRIPTION
**Pull request by asaunin**

### CODE REVIEW

File: src/test/java/part1/exercise/StreamsExercise1.java [hunk 1]
```diff
@@ -23,22 +22,50 @@
     // https://youtu.be/O8oN4KSZEXE Сергей Куксенко — Stream API, часть 1
     // https://youtu.be/i0Jr2l3jrDA Сергей Куксенко — Stream API, часть 2
 
+    private static List<Employee> employees;
+
+    private static final Predicate<String> employerWorkedForEpam = e -> e.equals("epam");
```
Comment (senia-psm) [inline src/test/java/part1/exercise/StreamsExercise1.java]: Use `"epam"::equals` instead

File: src/test/java/part1/exercise/StreamsExercise1.java [hunk 2]
```diff
@@ -23,22 +22,50 @@
...
+    private static final Consumer<Employee> printEmployee = e -> {
+        System.out.println(e.getPerson().toString());
+        e.getJobHistory().stream()
+                .peek(j -> System.out.println("\t\t".concat(j.toString())))
+                .map(getEmployer)
+                .anyMatch(employerWorkedForEpam);
+    };
+
+    @BeforeClass
+    public static void generateEmployeeList() {
```
Comment (senia-psm) [inline src/test/java/part1/exercise/StreamsExercise1.java]: Use static import instead

File: src/test/java/part1/exercise/StreamsExercise1.java [hunk 3]
```diff
@@ -49,11 +76,13 @@ public void sumEpamDurations() {
...
 
-        // TODO
-        throw new UnsupportedOperationException();
-
-        // int result = ???
-        // assertEquals(expected, result);
+        int result = employees.stream()
+                .map(Employee::getJobHistory)
+                .flatMap(Collection::stream)
+                .filter(j -> employerWorkedForEpam.test(getEmployer.apply(j)))
```
Comment (senia-psm) [inline src/test/java/part1/exercise/StreamsExercise1.java]: I guess `j.getEmployer().equals("epam")` is more readable.

File: src/test/java/part1/exercise/StreamsExercise2.java [hunk 4]
```diff
@@ -121,4 +156,78 @@ public void greatestExperiencePerEmployer() {
...
+    }
+
+    private static Stream<StreamsExercise2.PersonEmployerPair> firstEmployeeToPairs(Employee employee) {
+        return employee.getJobHistory().stream()
+                .map(JobHistoryEntry::getEmployer)
+                .limit(1)
+                .map(e -> new StreamsExercise2.PersonEmployerPair(employee.getPerson(), e));
+    }
+
+    private Map<String, Person> getGreatestExperiencePerEmployer(List<Employee> employees) {
```
Comment (senia-psm) [inline src/test/java/part1/exercise/StreamsExercise2.java]: You should prefer Person with 10 single year records over Person with a single 5 years record.

File: src/test/java/part1/exercise/StreamsExercise2.java [hunk 5]
```diff
@@ -121,4 +156,78 @@ public void greatestExperiencePerEmployer() {
...
+    @Test
+    public void employersStuffLists() {
+        System.out.println("Employers list stuff:");
+        Map<String, List<Person>> employersStuffLists = employees.stream()
+                .flatMap(StreamsExercise2::employeeToPairs)
+                .collect(Collectors.groupingBy(
+                        StreamsExercise2.PersonEmployerPair::getEmployer,
+                        mapping(StreamsExercise2.PersonEmployerPair::getPerson, toList())));
+
+        employersStuffLists.keySet()
```
Comment (senia-psm) [inline src/test/java/part1/exercise/StreamsExercise2.java]: Use asserts instead.

**Review by senia-psm — CHANGES_REQUESTED** [review]

File: src/test/java/part1/exercise/StreamsExercise1.java (hunk 3 above)
Comment (ASaunin) [inline src/test/java/part1/exercise/StreamsExercise1.java]: It's really less readable, but we already have employerWorkedForEpam function for compearing. 
I just wanted to use it

File: src/test/java/part1/exercise/StreamsExercise2.java (hunk 5 above)
Comment (ASaunin) [inline src/test/java/part1/exercise/StreamsExercise2.java]: Expected lists are too long, may I don't do this?
I implemented assertation in CollectorsExercise1 - it's rather short

Comment (senia-psm) [inline src/test/java/part1/exercise/StreamsExercise2.java]: As you like. Note that you can check not only equality, but also some other properties using `allMatch`, `nonMatch` etc.

File: src/test/java/part2/exercise/CollectorsExercise1.java [hunk 6]
```diff
@@ -51,32 +66,102 @@ public int getDuration() {
...
+        @Override
+        public Function<Map<String, PersonPositionDuration>, Map<String, Person>> finisher() {
+            return (m) -> m.entrySet().stream()
+                    .collect(toMap(Map.Entry::getKey, e -> e.getValue().getPerson()));
+        }
+
+        @Override
+        public Set<Characteristics> characteristics() {
+            return Collections.unmodifiableSet(
+                    EnumSet.of(Characteristics.CONCURRENT));
```
Comment (senia-psm) [inline src/test/java/part2/exercise/CollectorsExercise1.java]: Incorrect. `CONCURRENT` means that you can add elements into single accumulator from multiple threads. Actually you can't add elements into `HashMap` from multiple threads.

**Review by senia-psm — CHANGES_REQUESTED** [review]

File: src/test/java/part2/exercise/CollectorsExercise1.java (hunk 6 above)
Comment (ASaunin) [inline src/test/java/part2/exercise/CollectorsExercise1.java]: Can you explain one more time, how to set characteristics properly?

Comment (senia-psm) [inline src/test/java/part2/exercise/CollectorsExercise1.java]: Use empty set. There should be no properties.

**Pull request closed without merging**